File: revisao/dfs.py
class Vertice:
    def __init__(self):
        self.visitado = 0
        self.distancia = 0
        self.lista_adj = []

class Grafo:
    def __init__(self, qtd):
        self.qtd = qtd
        self.vertices = []
        for i in range(0, qtd):
            self.vertices.append(Vertice())

    def addAresta(self, v1, v2):
        if v1 >= self.qtd or v2 >= self.qtd or v1==v2:
            return False
        self.vertices[v1].lista_adj.append(v2)
        self.vertices[v2].lista_adj.append(v1)

    def addVertice(self):
        self.vertices.append(Vertice())
        self.qtd += 1

    def mostra(self):
        for i in range(len(self.vertices)):
            print(i, '->', self.vertices[i].lista_adj)
    
    def mostra_distancia(self, i):
        for j in range(len(self.vertices)):
            print('distancia de', i, 'para', j, ':', self.vertices[j].distancia)

    def dfs_distancia(self, i, distancia):
        self.vertices[i].visitado = 1
        self.vertices[i].distancia = distancia
        print(i)

        for j in range(0, len(self.vertices[i].lista_adj)):
            elemento = self.vertices[i].lista_adj[j]
            if self.vertices[ elemento ].visitado == 0:
                self.dfs_distancia(elemento, distancia+1)


grafo = Grafo(5)

grafo.addAresta(0, 1)
grafo.addAresta(0, 2)
grafo.addAresta(1, 3)
grafo.addAresta(1, 4)
grafo.addAresta(1, 5)
grafo.addAresta(4, 7)
grafo.addAresta(4, 4)

grafo.addVertice()

grafo.addAresta(4, 5)

grafo.dfs_distancia(0, 0)

grafo.mostra()

File: revisao/dfs.c
#include <stdio.h>
#include <stdlib.h>

typedef struct vertice {
    int visitado;
    int distancia;
    int cor;
    int lista_adj[100];
    int tamanho_lista;
}vertice;

int dfs_distancia(vertice * v, int i, int distancia);
int dfs_bipartido(vertice * v, int i, int cor);
int dfs_inout(vertice * v, int i, int * in, int * out, int * count);
int dfs_backedge(vertice * v, int i, int pai);

int main(){

    int n, a, x, y, i;

    scanf("%d", &n);

    vertice * v = malloc(sizeof(vertice)*(n+1));

    //for(i = 0; i < a; i++){
    while(1){
        scanf("%d %d", &x, &y);

        if(x == 0 || y == 0) break;
        v[x].lista_adj[ v[x].tamanho_lista++ ] = y;
        v[y].lista_adj[ v[y].tamanho_lista++ ] = x;
    }

    printf("\n%d", dfs_backedge(v, 1, 0));

    //IN OUT
    /* int count = 1, in[n+1], out[n+1];
    int el, pai;
    scanf("%d %d", &el, &pai);
    if(in[el] > in[pai] && out[el] < out[pai]){
        printf("\n%d é filho de %d", el, pai);
    }else{
        printf("\n%d NAO é filho de %d", el, pai);
    } */

    printf("\n");
    return 0;
}

int dfs_inout(vertice * v, int i, int * in, int * out, int * count)
{
    in[i] = *count;
    *count += 1;

    v[i].visitado = 1;

    for(int j = 0; j < v[i].tamanho_lista; j++){
        int elemento = v[i].lista_adj[j];
        if(v[ elemento ].visitado == 0){
            dfs_inout(v, elemento, in, out, count);
        }
    }


    out[i] = *count;
    *count += 1;
}

int dfs_bipartido(vertice * v, int i, int cor)
{

    v[i].visitado = 1;
    v[i].cor = cor;

    int bipartido = 1;

    for(int j = 0; j < v[i].tamanho_lista; j++){

        int elemento = v[i].lista_adj[j];
        if(v[ elemento ].visitado == 0){
            bipartido = dfs_bipartido(v, elemento, !cor);
        }else{
            if(v[ elemento ].cor == cor){
                bipartido = 0;
            }
        }
    }
    return bipartido;
}

int dfs_distancia(vertice * v, int i, int distancia)
{
    v[i].visitado = 1;
    v[i].distancia = distancia;

    for(int j = 0; j < v[i].tamanho_lista; j++){

        int elemento = v[i].lista_adj[j];
        if(v[ elemento ].visitado == 0){
            dfs_distancia(v, elemento, distancia+1);
        }
    }
}

int dfs_backedge(vertice * v, int i, int pai)
{
    v[i].visitado = 1;

    int hasBE = 0;

    for(int j = 0; j < v[i].tamanho_lista; j++){
        int elemento = v[i].lista_adj[j];

        if(v[ elemento ].visitado == 0){
            hasBE = dfs_backedge(v, elemento, i);
        }else{
            if(pai != elemento){
                hasBE = 1;
            }
        }
    }

    return hasBE;
}
